aten/src/ATen/cuda/CUDAGraph.h

```diff
@@ -18,22 +18,32 @@ namespace cuda {
 // to CUDAGraph::capture_begin
 TORCH_CUDA_CPP_API MempoolId_t graph_pool_handle();
 
-void dynamic_graph_updater(cudaGraphKernelNodeUpdate* updates, size_t num_updates);
-// Large parameter support:
-// https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
-// Question: Does pytorch still run on Pascal or older?
 
 struct KernelUpdateSOA {
-  static constexpr size_t MAX_NUM_UPDATES =
-      1364; // (32764 - sizeof(size_t)) / (sizeof(void*) + sizeof(size_t) +
-            // sizeof(cudaGraphDeviceNode_t));
+  // Even on Cuda 12.4, Torch supports compute capability down to sm_50
+  // However, large parameter support was only added in sm_70
+  // https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
+#if (__CUDA_ARCH__ >= 700)
```

Comment:
This is a good idea, but it is error prone unfortunately. We currently reference this struct in CUDAGraph.cpp and CUDAGraph.cu (and potentially every files that may include CUDAGraph.h). What will happen is that the host code will set KERNEL_PARAM_LIMIT_BYTES to 4096, which is fine for Pascal and older, but will cause problems for Volta and newer code, since the host and device layouts of this type won't match. AFAIK, there is no way for the host compiler to "know" for which devices the same code is being compiled.

Anyway, honestly just using KERNEL_PARAM_LIMIT_BYTES=4096 may be fine. Looks like that gives us 73 updates. I would like to get stats on that when running torchbench, but my hunch is that 73 is enough. (And of course, we can call the kernel multiple times, as you are doing).

Anyway, it looks like we can do 73 updates in a cross-platform way. That is probably okay for now TBH.

Reply:
I believe we can use templates to have it both ways, and pick at runtime: https://github.com/pytorch/pytorch/pull/137318/commits/518e270bf1dc3924f5cb9fbf2acb161be3d76995

73 is low, especially in the scenario where the private pool is discarded and all allocations are rewritten (not just inputs/outputs)